hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/UpsertProcessorSupplier.java

```diff
@@ -0,0 +1,143 @@
+
+import com.hazelcast.jet.core.Processor;
+import com.hazelcast.jet.impl.connector.WriteJdbcP;
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.DataSerializable;
+import com.hazelcast.security.impl.function.SecuredFunction;
+import com.hazelcast.security.permission.ConnectorPermission;
+import com.hazelcast.sql.impl.row.JetSqlRow;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.security.Permission;
+import java.sql.PreparedStatement;
+import java.sql.SQLException;
+import java.sql.SQLNonTransientException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+import static com.hazelcast.security.permission.ActionConstants.ACTION_WRITE;
+import static java.util.Collections.singletonList;
+
+public class UpsertProcessorSupplier
+        extends AbstractJdbcSqlConnectorProcessorSupplier
+        implements DataSerializable, SecuredFunction {
+
+    private String query;
+    private int batchLimit;
+
+    // No-arg constructor for DataSerializable
+    @SuppressWarnings("unused")
+    public UpsertProcessorSupplier() {
+    }
+
+    public UpsertProcessorSupplier(String externalDataStoreRef, String query, int batchLimit) {
+        super(externalDataStoreRef);
+        this.query = query;
+        this.batchLimit = batchLimit;
+    }
+
+    @Nonnull
+    @Override
+    public Collection<? extends Processor> get(int count) {
+        // count means local parallelism. Create as many processors as required
+        List<Processor> processors = new ArrayList<>(count);
+        for (int i = 0; i < count; i++) {
+            Processor processor = new WriteJdbcP<>(
+                    query,
+                    dataSource,
+                    (PreparedStatement ps, JetSqlRow row) -> {
+                        for (int j = 0; j < row.getFieldCount(); j++) {
+                            // JDBC parameterIndex is 1-based, so j + 1
+                            ps.setObject(j + 1, row.get(j));
+                        }
+                    },
+                    this::isNonTransientException,
+                    false,
+                    batchLimit
+            );
+            processors.add(processor);
+        }
+        return processors;
+    }
+
+    @SuppressWarnings("BooleanExpressionComplexity")
+    private boolean isNonTransientException(SQLException e) {
+        SQLException next = e.getNextException();
+        return e instanceof SQLNonTransientException
+                || e.getCause() instanceof SQLNonTransientException
+                || !isTransientCode(e.getSQLState())
+                || (next != null && e != next && isNonTransientException(next));
+    }
+
+    private boolean isTransientCode(String code) {
+        // Full list of error codes at:
+        // https://www.postgresql.org/docs/current/errcodes-appendix.html
```

Comment:
Isn't this postgres-specific? Are we using it for all databases?

Reply:
It was an example to a link. I have deleted the comment

Reply:
According to  @Frantisek it is postgre specific because postgre does not throw SQLNonTransientException but only throws org.postgresql.util.PSQLException